pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,550 @@
+	vmList, err := Esxclient.Projects.GetVMs(project, nil)
+	if err != nil {
+		return "", fmt.Errorf("Failed to GetVMs for project %s, error [%v]", project, err)
+	}
+
+	for _, vm := range vmList.Items {
+		task, err := Esxclient.VMs.GetNetworks(vm.ID)
+		if err != nil {
+			glog.Warningf("Photon Cloud Provider: GetNetworks failed for vm.ID %s, error [%v]", vm.ID, err)
+		} else {
+			task, err = Esxclient.Tasks.Wait(task.ID)
+			if err != nil {
+				glog.Warning("Photon Cloud Provider: Wait task for GetNetworks failed for vm.ID %s, error [%v]", vm.ID, err)
+			} else {
+				networkConnections := task.ResourceProperties.(map[string]interface{})
+				networks := networkConnections["networkConnections"].([]interface{})
+				for _, nt := range networks {
+					network := nt.(map[string]interface{})
+					if val, ok := network["ipAddress"]; ok && val != nil {
+						ipAddr := val.(string)
+						if ipAddr == IPAddress {
+							return vm.ID, nil
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("No matching VM is found with IP %s", IPAddress)
+}
+
+// Retrieve the the Photon project ID from the Photon Controller endpoint based on the project name
+func getProjIDbyName(tenantName, projName string) (string, error) {
+	tenants, err := Esxclient.Tenants.GetAll()
+	if err != nil {
+		return "", fmt.Errorf("GetAll tenants failed with error [%v]", err)
+	}
+
+	for _, tenant := range tenants.Items {
+		if tenant.Name == tenantName {
+			projects, err := Esxclient.Tenants.GetProjects(tenant.ID, nil)
+			if err != nil {
+				return "", fmt.Errorf("Failed to GetProjects for tenant %s, error [%v]", tenantName, err)
+			}
+
+			for _, project := range projects.Items {
+				if project.Name == projName {
+					return project.ID, nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("No matching tenant/project name is found with %s/%s", tenantName, projName)
+}
+
+func newPCCloud(cfg PCConfig) (*PCCloud, error) {
+	if len(cfg.Global.CloudTarget) == 0 {
+		return nil, fmt.Errorf("Invalid Photon Controller endpoint")
+	}
+
+	//TODO: add handling of certification enabled situation
+	options := &photon.ClientOptions{
+		IgnoreCertificate: cfg.Global.IgnoreCertificate,
+	}
+
+	Esxclient = photon.NewClient(cfg.Global.CloudTarget, options, logger)
+	status, err := Esxclient.Status.Get()
+	if err != nil {
+		glog.Errorf("Photon Cloud Provider: new client creation failed, error: [%v], status: [%v]", err, status)
+		return nil, logError("new client creation failed", err)
+	}
+	glog.V(2).Info("Photon Cloud Provider: Status of the new photon controller client: %v", status)
+
+	// Get Photon Controller project ID for future use
+	projID, err := getProjIDbyName(cfg.Global.Tenant, cfg.Global.Project)
+	if err != nil {
+		return nil, logError("getProjIDbyName failed when creating new Photon Controller client", err)
+	}
+
+	// Get local hostname for localInstanceID
+	cmd := exec.Command("bash", "-c", `echo $HOSTNAME`)
```

Comment:
I think this should be the same reliable level as hostname command... open to suggestions.